src/main/java/org/tugraz/sysds/runtime/lineage/LineageCache.java

```diff
@@ -1,60 +1,307 @@
+/*
+ * Copyright 2019 Graz University of Technology
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package org.tugraz.sysds.runtime.lineage;
 
 import org.tugraz.sysds.api.DMLScript;
+import org.tugraz.sysds.hops.OptimizerUtils;
+import org.tugraz.sysds.hops.cost.CostEstimatorStaticRuntime;
+import org.tugraz.sysds.lops.MMTSJ.MMTSJType;
+import org.tugraz.sysds.runtime.DMLRuntimeException;
 import org.tugraz.sysds.runtime.controlprogram.caching.MatrixObject;
 import org.tugraz.sysds.runtime.controlprogram.context.ExecutionContext;
+import org.tugraz.sysds.runtime.instructions.CPInstructionParser;
 import org.tugraz.sysds.runtime.instructions.Instruction;
+import org.tugraz.sysds.runtime.instructions.cp.CPInstruction.CPType;
 import org.tugraz.sysds.runtime.instructions.cp.ComputationCPInstruction;
+import org.tugraz.sysds.runtime.instructions.cp.MMTSJCPInstruction;
 import org.tugraz.sysds.runtime.matrix.data.MatrixBlock;
+import org.tugraz.sysds.runtime.util.LocalFileUtils;
 
+import java.io.IOException;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Map;
 
 public class LineageCache {
-	private static final Map<LineageItem, MatrixBlock> _cache = new HashMap<>();
-	
+	private static final Map<LineageItem, Entry> _cache = new HashMap<>();
```

Comment:
Consider using a `LinkedHashMap` here to avoid the custom Entry implementation. I guess you could emulate LRU by doing another put which should place the object at the end of the queue.

Reply:
As I realized, `LinkedHashMap` is a cleaner replacement of the custom data structure, but I have a feeling that I might need to use the LinkedList nodes to store few more things.  If that doesn't turn out to be the case, I will replace in next couple of PRs.